File: icebreaker_generator.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const { createReadStream, createWriteStream } = require('fs');
const { parse } = require('csv-parse');
const { stringify } = require('csv-stringify');

// Load environment variables
try {
  require('dotenv').config({ path: path.join(__dirname, '.env') });
} catch (e) {
  // dotenv not available, continue with environment variables
}

// Configure paths
const CSV_PATH = 'C:\\Users\\79818\\Downloads\\call centers US UK Aus 10-100 - 10-50.csv';
const OUTPUT_DIR = path.join(__dirname, 'data', 'processed');
const OUTPUT_FILE = path.join(OUTPUT_DIR, `apollo_icebreaker_analyzed_${Date.now()}.csv`);
const LOG_FILE = path.join(OUTPUT_DIR, 'generator.log');

// Ensure directories exist
fs.mkdirSync(OUTPUT_DIR, { recursive: true });

// Logging
const logStream = fs.createWriteStream(LOG_FILE, { encoding: 'utf-8' });

function log(msg) {
  console.log(msg);
  logStream.write(msg + '\n');
}

log('='.repeat(60));
log('ICEBREAKER GENERATOR (Node.js)');
log('='.repeat(60));

// Import Anthropic
let Anthropic;
try {
  Anthropic = require('@anthropic-ai/sdk');
  log('[OK] Anthropic SDK loaded');
} catch (e) {
  log(`[FAIL] Anthropic SDK not found: ${e.message}`);
  log('Installing via npm...');
  const { execSync } = require('child_process');
  try {
    execSync('npm install @anthropic-ai/sdk', { stdio: 'inherit' });
    Anthropic = require('@anthropic-ai/sdk');
    log('[OK] Anthropic SDK installed');
  } catch (e2) {
    log(`[FAIL] Could not install: ${e2.message}`);
    process.exit(1);
  }
}

// Initialize client
const apiKey = process.env.ANTHROPIC_API_KEY;
if (!apiKey) {
  log('[FAIL] ANTHROPIC_API_KEY not set');
  log(`Available env vars: ${Object.keys(process.env).filter(k => k.includes('ANTHROPIC')).join(', ')}`);
  process.exit(1);
}

const client = new Anthropic({ apiKey });
log('[OK] Anthropic client initialized');

async function generateIcebreaker(fullName, companyName, title, headline, city) {
  const prompt = `You are an outreach message generator.
Your role: create short, casual, human-sounding icebreaker messages for LinkedIn-style outreach.
Goal: make the recipient feel recognized for their work without sounding pushy or overly formal.

Your task:
- If "full_name" looks like a company (contains words such as Company, Inc, LLC, Ltd, Group, Realty, Properties, Brokers, UAE, Dubai, Abu Dhabi):
    -> Output ONLY: its a company

- Else (if "full_name" is a person):
    1. Extract firstName = first word of full_name.
    2. Normalize company_name into shortCompany:
        - If ALL CAPS -> convert to Title Case (only first letter uppercase).
        - Remove corporate suffixes: Properties, Realty, Group, Brokers, LLC, Ltd, Inc, UAE, Dubai, Abu Dhabi.
        - Remove apostrophes or special symbols.
    3. Normalize region into shortRegion:
        - Dubai, Abu Dhabi -> Dubai
        - San Francisco -> SF
        - New York City -> NYC
        - Else keep original.
    4. Generate opening = pick randomly, in casual tone:
        - love how you
        - really like how you
        - awesome to see you
        - impressed by how you
        - great track with how you
        - cool to see you
    5. specializationPhrase:
        - Look at headline or title.
        - If clear keyword (luxury, sales, marketing, engineering, talent acquisition, product, design, etc.) -> rewrite naturally as an action (2-3 words).
            * Example: "Luxury Consultant" -> "drive luxury sales"
            * "Marketing Manager" -> "lead marketing"
            * "Talent Acquisition" -> "grow teams"
            * "Software Engineer" -> "build products"
        - If generic title -> simplify to meaningful action:
            * "Consultant" -> "work with clients"
            * "Broker" -> "push sales"
            * "Analyst" -> "dig into insights"
        - If nothing useful -> fallback: "bring industry experience".
    6. regionPhrase = pick randomly:
        - I'm also in the {shortRegion} market
        - I work across {shortRegion} as well
        - I'm active in {shortRegion} too
        - I also focus on {shortRegion}
    7. closingPhrase = pick randomly:
        - Wanted to run something by you.
        - Thought I'd share an idea with you.
        - Had something you might find interesting.
        - Figured I'd reach out quickly.

Final Output (always one line, no labels, no JSON):
Hey {firstName}, {opening} {specializationPhrase} at {shortCompany} - {regionPhrase}. {closingPhrase}

Context for this row:
full_name: ${fullName}
company_name: ${companyName}
title: ${title}
headline: ${headline}
city: ${city}`;

  try {
    const response = await client.messages.create({
      model: 'claude-haiku-4-5-20251001',
      max_tokens: 200,
      messages: [
        {
          role: 'user',
          content: prompt
        }
      ]
    });

    return response.content[0].text.trim();
  } catch (e) {
    return `ERROR: ${e.message.substring(0, 40)}`;
  }
}

async function main() {
  log('[1/3] Loading CSV...');

  const rows = [];
  const errorHandler = (err) => {
    log(`[ERROR] CSV Parse: ${err.message}`);
    process.exit(1);
  };

  await new Promise((resolve, reject) => {
    createReadStream(CSV_PATH, { encoding: 'utf-8' })
      .pipe(parse({ columns: true }))
      .on('data', (row) => rows.push(row))
      .on('error', errorHandler)
      .on('end', resolve);
  });

  log(`[OK] Loaded ${rows.length} rows`);

  // Process rows in batches
  log('\n[2/3] Generating icebreakers...');
  const BATCH_SIZE = 50;
  const startTime = Date.now();

  let successCount = 0;
  let companyCount = 0;
  let errorCount = 0;

  for (let i = 0; i < rows.length; i += BATCH_SIZE) {
    const batchNum = Math.floor(i / BATCH_SIZE) + 1;
    const batchEnd = Math.min(i + BATCH_SIZE, rows.length);
    const batchSize = batchEnd - i;

    log(`\n[BATCH ${batchNum}] Processing rows ${i + 1}-${batchEnd}...`);
    const batchStart = Date.now();

    // Process batch in parallel
    const batchPromises = [];
    for (let j = i; j < batchEnd; j++) {
      const row = rows[j];
      const promise = generateIcebreaker(
        row.full_name || '',
        row.company_name || '',
        row.title || '',
        row.headline || '',
        row.city || ''
      ).then((icebreaker) => {
        rows[j].icebreaker = icebreaker;

        if (icebreaker.includes('ERROR')) {
          errorCount++;
        } else if (icebreaker.toLowerCase().includes('its a company')) {
          companyCount++;
        } else {
          successCount++;
        }

        return icebreaker;
      });

      batchPromises.push(promise);
    }

    await Promise.all(batchPromises);

    const batchTime = ((Date.now() - batchStart) / 1000).toFixed(1);
    log(`  Completed in ${batchTime}s | Success: ${successCount}, Companies: ${companyCount}, Errors: ${errorCount}`);
  }

  const totalTime = ((Date.now() - startTime) / 1000).toFixed(1);
  log(`\n[OK] Generation complete in ${totalTime}s`);

  // Save CSV
  log('\n[3/3] Saving results...');

  const csvOutput = createWriteStream(OUTPUT_FILE);
  const stringifier = stringify({ header: true, columns: Object.keys(rows[0]) });

  stringifier.pipe(csvOutput);

  for (const row of rows) {
    stringifier.write(row);
  }

  stringifier.end();

  await new Promise((resolve, reject) => {
    csvOutput.on('finish', resolve);
    csvOutput.on('error', reject);
  });

  log(`[OK] Saved to ${OUTPUT_FILE}`);

  // Stats
  log('\n' + '='.repeat(60));
  log('FINAL STATISTICS');
  log('='.repeat(60));
  log(`Total rows: ${rows.length}`);
  log(`Successful: ${successCount}`);
  log(`Companies detected: ${companyCount}`);
  log(`Errors: ${errorCount}`);
  log(`Total time: ${totalTime}s`);
  log(`Output: ${OUTPUT_FILE}`);
  log('='.repeat(60));

  // Sample
  log('\nSAMPLE ICEBREAKERS (first 10):');
  let sampleCount = 0;
  for (const row of rows) {
    if (!row.icebreaker.includes('ERROR') && !row.icebreaker.toLowerCase().includes('its a company')) {
      sampleCount++;
      log(`\n[${sampleCount}] ${row.full_name} @ ${row.company_name}`);
      log(`    ${row.icebreaker}`);
      if (sampleCount >= 10) break;
    }
  }

  logStream.end();
  process.exit(0);
}

main().catch((err) => {
  log(`FATAL ERROR: ${err.message}`);
  logStream.end();
  process.exit(1);
});

File: modules/apollo/icebreaker_generator.py
#!/usr/bin/env python3
"""
=== ICEBREAKER GENERATOR ===
Version: 1.0.0 | Created: 2025-11-03

PURPOSE:
Generate personalized icebreaker messages for Apollo CSV leads

FEATURES:
- Company vs person detection
- Smart name/company/region normalization
- Random variation for natural messages
- Context-aware specialization phrases

USAGE:
1. Configure CSV_PATH in CONFIG
2. Run: python icebreaker_generator.py
3. Results saved to data/processed/
"""

import pandas as pd
import random
import re
from pathlib import Path
from datetime import datetime


class SimpleLogger:
    """Simple console logger"""
    @staticmethod
    def info(msg):
        print(f"[INFO] {msg}")

    @staticmethod
    def error(msg, exc_info=False):
        print(f"[ERROR] {msg}")


logger = SimpleLogger()

CONFIG = {
    "CSV_PATH": r"C:\Users\79818\Downloads\call centers US UK Aus 10-100 - 10-50.csv",
    "OUTPUT_DIR": Path(__file__).parent.parent.parent / "data" / "processed",
    "BATCH_SIZE": 200
}

# Opening variations
OPENINGS = [
    "love how you",
    "really like how you",
    "awesome to see you",
    "impressed by how you",
    "great track with how you",
    "cool to see you"
]

# Region phrase templates
REGION_PHRASES = [
    "I'm also in the {region} market",
    "I work across {region} as well",
    "I'm active in {region} too",
    "I also focus on {region}"
]

# Closing phrases
CLOSINGS = [
    "Wanted to run something by you.",
    "Thought I'd share an idea with you.",
    "Had something you might find interesting.",
    "Figured I'd reach out quickly."
]

# Company identifiers
COMPANY_KEYWORDS = [
    "Company", "Inc", "LLC", "Ltd", "Group", "Realty",
    "Properties", "Brokers", "UAE", "Dubai", "Abu Dhabi"
]

# Suffixes to remove from company names
COMPANY_SUFFIXES = [
    "Properties", "Realty", "Group", "Brokers", "LLC",
    "Ltd", "Inc", "UAE", "Dubai", "Abu Dhabi", "Limited",
    "Corporation", "Corp", "Co", "LLP", "PLC"
]

# Specialization keywords and actions
SPECIALIZATIONS = {
    "luxury": "drive luxury sales",
    "marketing": "lead marketing",
    "sales": "push sales",
    "engineering": "build products",
    "engineer": "build products",
    "talent": "grow teams",
    "product": "build products",
    "design": "shape design",
    "consultant": "work with clients",
    "consulting": "work with clients",
    "broker": "push sales",
    "analyst": "dig into insights",
    "executive": "lead strategy",
    "founder": "build companies",
    "ceo": "lead strategy",
    "cto": "drive tech",
    "cfo": "manage growth",
    "coo": "run operations",
    "director": "drive results",
    "manager": "lead teams",
    "specialist": "bring expertise",
    "coordinator": "organize growth"
}


def is_company(full_name: str) -> bool:
    """Check if full_name represents a company rather than a person"""
    if pd.isna(full_name):
        return False

    return any(keyword.lower() in full_name.lower() for keyword in COMPANY_KEYWORDS)


def normalize_company(company_name: str) -> str:
    """Normalize company name: Title Case + remove suffixes"""
    if pd.isna(company_name):
        return ""

    # Convert all-caps to Title Case
    if company_name.isupper():
        company_name = company_name.title()

    # Remove suffixes
    for suffix in COMPANY_SUFFIXES:
        # Remove with comma and space
        company_name = re.sub(rf',\s*{re.escape(suffix)}\b', '', company_name, flags=re.IGNORECASE)
        # Remove without comma
        company_name = re.sub(rf'\s+{re.escape(suffix)}\b', '', company_name, flags=re.IGNORECASE)

    # Remove apostrophes and special characters
    company_name = re.sub(r"['\"]", '', company_name)

    # Clean up extra spaces
    company_name = ' '.join(company_name.split())

    return company_name.strip()


def normalize_region(region: str) -> str:
    """Normalize region names to short forms"""
    if pd.isna(region):
        return ""

    region = region.strip()

    # Region mappings
    mappings = {
        "Dubai": "Dubai",
        "Abu Dhabi": "Dubai",
        "San Francisco": "SF",
        "New York City": "NYC",
        "New York": "NYC"
    }

    return mappings.get(region, region)


def extract_specialization(headline: str, title: str) -> str:
    """Extract specialization phrase from headline or title"""
    text = f"{headline} {title}".lower()

    # Check for specialization keywords
    for keyword, action in SPECIALIZATIONS.items():
        if keyword in text:
            return action

    # Default fallback
    return "bring industry experience"


def generate_icebreaker(row: dict) -> str:
    """Generate personalized icebreaker for a row"""
    full_name = row.get('full_name', '')

    # Check if company
    if is_company(full_name):
        return "its a company"

    # Extract first name
    first_name = full_name.split()[0] if full_name else "there"

    # Normalize company
    company_name = row.get('company_name', '')
    short_company = normalize_company(company_name)

    # Normalize region (city or state)
    city = row.get('city', '')
    state = row.get('state', '')
    short_region = normalize_region(city if city else state)

    # Extract specialization
    headline = row.get('headline', '')
    title = row.get('title', '')
    specialization_phrase = extract_specialization(headline, title)

    # Random selections
    opening = random.choice(OPENINGS)
    region_template = random.choice(REGION_PHRASES)
    closing = random.choice(CLOSINGS)

    # Build region phrase
    region_phrase = region_template.format(region=short_region) if short_region else "I'm also in the area"

    # Build final icebreaker
    icebreaker = f"Hey {first_name}, {opening} {specialization_phrase} at {short_company} - {region_phrase}. {closing}"

    return icebreaker


def process_batch(df_batch: pd.DataFrame, batch_num: int, total_batches: int) -> pd.DataFrame:
    """Process a batch of rows"""
    logger.info(f"Processing batch {batch_num}/{total_batches} ({len(df_batch)} rows)")

    # Generate icebreakers
    df_batch['icebreaker'] = df_batch.apply(
        lambda row: generate_icebreaker(row.to_dict()),
        axis=1
    )

    # Count companies vs persons
    companies = (df_batch['icebreaker'] == "its a company").sum()
    persons = len(df_batch) - companies

    logger.info(f"Batch {batch_num} complete: {persons} persons, {companies} companies")

    return df_batch


def main():
    logger.info("=== ICEBREAKER GENERATOR START ===")

    try:
        # Read CSV
        csv_path = Path(CONFIG["CSV_PATH"])
        if not csv_path.exists():
            raise FileNotFoundError(f"CSV file not found: {csv_path}")

        logger.info(f"Reading CSV: {csv_path}")
        df = pd.read_csv(csv_path)

        total_rows = len(df)
        logger.info(f"Total rows: {total_rows}")

        # Calculate batches
        batch_size = CONFIG["BATCH_SIZE"]
        total_batches = (total_rows + batch_size - 1) // batch_size
        logger.info(f"Processing in {total_batches} batches of {batch_size}")

        # Process in batches
        results = []
        for i in range(total_batches):
            start_idx = i * batch_size
            end_idx = min((i + 1) * batch_size, total_rows)

            batch = df.iloc[start_idx:end_idx].copy()
            processed_batch = process_batch(batch, i + 1, total_batches)
            results.append(processed_batch)

        # Combine all batches
        logger.info("Merging all batches...")
        df_final = pd.concat(results, ignore_index=True)

        # Statistics
        total_companies = (df_final['icebreaker'] == "its a company").sum()
        total_persons = len(df_final) - total_companies

        logger.info("=== FINAL STATISTICS ===")
        logger.info(f"Total processed: {len(df_final)}")
        logger.info(f"Persons: {total_persons}")
        logger.info(f"Companies: {total_companies}")

        # Sample icebreakers (persons only)
        logger.info("=== SAMPLE ICEBREAKERS ===")
        samples = df_final[df_final['icebreaker'] != "its a company"]['icebreaker'].head(10)
        for idx, sample in enumerate(samples, 1):
            logger.info(f"{idx}. {sample}")

        # Save results
        output_dir = CONFIG["OUTPUT_DIR"]
        output_dir.mkdir(parents=True, exist_ok=True)

        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        output_file = output_dir / f"apollo_icebreakers_{timestamp}.csv"

        df_final.to_csv(output_file, index=False)
        logger.info(f"Results saved to: {output_file}")

        logger.info("=== ICEBREAKER GENERATOR COMPLETE ===")

    except Exception as e:
        logger.error(f"Script failed: {e}", exc_info=True)
        raise


if __name__ == "__main__":
    main()
